pkg/controllers/chainer_controller.go

```diff
@@ -0,0 +1,461 @@
+
+package controllers
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+	appsv1 "k8s.io/api/apps/v1"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	appsinformers "k8s.io/client-go/informers/apps/v1"
+	batchinformers "k8s.io/client-go/informers/batch/v1"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	rbacinformers "k8s.io/client-go/informers/rbac/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	typedcorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+
+	apisv1alpha1 "github.com/kubeflow/chainer-operator/pkg/apis/chainer/v1alpha1"
+	clientset "github.com/kubeflow/chainer-operator/pkg/client/clientset/versioned"
+	kubeflowScheme "github.com/kubeflow/chainer-operator/pkg/client/clientset/versioned/scheme"
+	informers "github.com/kubeflow/chainer-operator/pkg/client/informers/externalversions/chainer/v1alpha1"
+	listers "github.com/kubeflow/chainer-operator/pkg/client/listers/chainer/v1alpha1"
+	"github.com/kubeflow/chainer-operator/pkg/controllers/backends"
+	"github.com/kubeflow/chainer-operator/pkg/controllers/backends/mpi"
+	"github.com/kubeflow/chainer-operator/pkg/controllers/backends/none"
+)
+
+const (
+	controllerAgentName = "chainer-job-controller"
+
+	// event types
+	failedSynced  = "FailedSynced"
+	successSynced = "Synced"
+
+	// event messages
+	messageBackendTypeNotSupported = "backend:%q not supported"
+	messageResourceSynced          = "ChainerJob synced successfully"
+
+	noneBackendKey = "__NONE__"
+)
+
+// ChainerJobController is the controller
+type ChainerJobController struct {
+	// kubeClient is a standard kubernetes clientset.
+	kubeClient kubernetes.Interface
+	// kubeflowClient is a clientset for our own API group.
+	kubeflowClient clientset.Interface
+
+	configMapSynced      cache.InformerSynced
+	serviceAccountSynced cache.InformerSynced
+	roleSynced           cache.InformerSynced
+	roleBindingSynced    cache.InformerSynced
+	statefulSetSynced    cache.InformerSynced
+	jobSynced            cache.InformerSynced
+	chainerJobLister     listers.ChainerJobLister
+	chainerJobSynced     cache.InformerSynced
+
+	// queue is a rate limited work queue. This is used to queue work to be
+	// processed instead of performing it as soon as a change happens. This
+	// means we can ensure we only process a fixed amount of resources at a
+	// time, and makes it easy to ensure we are never processing the same item
+	// simultaneously in two different workers.
+	queue workqueue.RateLimitingInterface
+
+	// recorder is an event recorder for recording Event resources to the
+	// Kubernetes API.
+	recorder record.EventRecorder
+
+	backends map[apisv1alpha1.BackendType]backends.Backend
```

Comment:
Personally, I do not recommend maintaining such a cache in memory, and k8s community also does not recommend. It will be better to be event-driven. While LGTM since this is the first commit. We could merge it and go ahead iteratively.

Reply:
Thanks!!  I will.  I opened the issue on it https://github.com/kubeflow/chainer-operator/issues/13